contracts/InverseVester.sol

```diff
@@ -0,0 +1,114 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.3;
+import "@openzeppelin/contracts/access/Ownable.sol";
+import "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import "./interfaces/IInv.sol";
+
+/**
+ * @title Inverse token vesting contract
+ * @author Inverse Finance
+ * @notice Base contract for vesting agreement on INV tokens
+ *         Reverse vesting means tokens are owned but the recipient
+ *         and can be used for voting.
+ * @dev    Vesting calculation is linear
+ */
+contract InverseVester is Ownable {
+    using SafeERC20 for IInv;
+
+    uint256 public constant DAY = 1 days;
+
+    // @dev Total amount to be vested
+    uint256 public vestingAmount;
+
+    // @dev Timestamp for the start of this vesting agreement
+    uint256 public vestingBegin;
+
+    // @dev Timestamp for the end of this vesting agreement
+    uint256 public vestingEnd;
+
+    // @dev Timestamp for the last time vested tokens were claimed
+    uint256 public lastClaimTimestamp;
+
+    // @dev Inverse finance treasury token
+    IInv public inv;
+
+    // @dev Amount of days the vesting period will last
+    uint16 public vestingDurationInDays;
+
+    // @dev Whether this is a reverse vesting agreement
+    bool public reverseVesting;
+
+    constructor(
```

Comment:
What do you think of transferring ownership to the treasury at contract creation, and have the treasury call `start`? Looks like the action of starting the vesting contract currently resides at the side of the recipient, whereas I would expect it to be at the sending side.

Reply:
Not sure I understand. My assumption is that treasury will call deploy and call start.

Reply:
The treasury is actually a timelock contract: https://etherscan.io/address/0x926df14a23be491164dcf93f4c468a50ef659d5b

While it can run arbitrary functions, it doesn't have anything built in to deploy arbitrary contracts. So in the current setup the process would be something like this;
 1. Person X deploys the vesting contract. Person X is now owner of the vesting contract.
 2. Delegate shoots in a proposal to call `inv.approve(<vesting-contract-address>, <vesting-amount>)`. This takes 3 days of voting.
 3. Someone calls `queue` after proposal passes. This takes 2 days.
 4. Someone calls `execute` after the 2 days. Proposal is executed, which means the action defined in step 2 is executed.
 5. Person X calls `start(<recipient-address>, <treasury-address>)`. The treasury is now the owner of the vesting contract.
 6. Vesting begins.

This does not seem ideal to me. I'd rather see one of these scenarios:
 - The vesting contract constructor accepts a treasury address and immediately transfers ownership to the treasury after it creates the contract. This way only the treasury calls `start`, which means the treasury address can be removed from the `start` function signature. It also means that steps 2 and 5 above can be combined so that vesting immediately starts after the proposal is executed at step 4.
 - A factory contract is added to create vesting contracts. This way the treasury can just call a function on that factory contract to create new vesting contracts with all the values being set at the time of proposal.

Reply:
In the current implementation the steps should be:

1. InvDeployer deploys contract
2. InvDeployer calls contract.transferOwnership(timelock)
3. Governance process...
4. Timelock executes inv.approve(contract, amount)
5. Timelock executes contract.start(recipient, address(this))

Am I missing something obvious?

Reply:
The Inverse Finance Deployer account is (as far as I know) just one of Nour's wallets. Preferably I'd like to decouple Nour from the process; he's busy enough.